#22 Fix Uncaught rejected promise

Rayhatron opened the pull request:
Disable parse in apiFetch requests and handle the response object to avoid errors where some other plugins can have a middleware which disables apiFetch parse and therefore results in this plugin throwing an uncaught error.

An example of such a third-party plugin with that behaviour is Yoast SEO Premium which has a createRedirectMiddleware on apiFetch that explicitly intercepts the apiFetch options and changes parse to false because it needs the response object (and not the parsed body of the response) to be able to check its headers and do some extra action if a specific Yoast header is found. Yoast SEO Premium version 18.5 has comments that document exactly what this paragraph says in a file titled "assets/js/dist/wp-seo-premium-redirect-notifications-1850.min.js" (recent versions seem to have those comments removed).

The Yoast code does return the parsed body of the response but if its middleware is executed first then the last middleware which in investigations could be the default WordPress ones will attempt to parse the response body but first checks if parse is true in options. Since it was set to false then the response object will be returned see [here](https://github.com/WordPress/gutenberg/blob/57fe11888c7939cd751ea3107675936e448fe760/packages/api-fetch/src/utils/response.js#L74).

To avoid such behaviour of third-party plugin middleware breaking this plugin, the plugin can just expect the response object in the apiFetch calls and then parses the body of the response.

The proposed solution has been implemented in this PR. #22 initiated the investigation of this.

kadamwhite reviewed (commented):
Makes sense but I haven't looked at this in a while, so two questions: 
- are there any other scenarios where we may have written other code which would be impacted by switching this parse statement?
- could we have addressed this by sniffing for the presence of response.json in the handler function, instead of manually disabling parse? (I'm thinking of, whether any other middleware might try to RE-enable parsing)

Rayhatron commented:
@kadamwhite I've updated the code to handle both response types since as you said, it is indeed possible that another middleware could reenable the parse option and the earlier code would start failing. Given the new changes, I've reverted the parse option to be left as true as it was originally in the code.

kadamwhite commented on assets/src/components/with-term/index.js:
```diff
@@ -33,30 +33,66 @@ export default function withTerm() {
+					} );
+				}
+
+				function handeNewTerm( newTerm ) {
```
Typo, `handle`

kadamwhite replied:
Nitpick, it'd be nice to have a basic JSDoc block above these handler functions. Non-essential.

kadamwhite commented on assets/src/components/with-term/index.js:
```diff
@@ -33,30 +33,66 @@ export default function withTerm() {
-							slug,
-						} );
+						if ( response instanceof Response ) {
+							response.json().then(
+								newTerm => {
+									handeNewTerm( newTerm );
```
```js
							response.json().then( handleNewTerm );
```
Since the callback does nothing but pass the argument to the handler, the handler could be used itself in place of the callback

kadamwhite commented on assets/src/components/with-term/index.js:
```diff
@@ -33,30 +33,66 @@ export default function withTerm() {
+				createTerm( taxonomy, label, value )
+					.catch( response => {
+						if ( response instanceof Response ) {
+							response.json().then(
+								error => {
+									handleError( error );
```
```suggestion
							response.json().then( handleError );
```
Same suggestion -- handler can become the callback

kadamwhite commented on assets/src/components/with-term/index.js:
```diff
@@ -33,30 +33,66 @@ export default function withTerm() {
+						if ( searchTermsResponse instanceof Response ) {
+							searchTermsResponse.json().then(
+								terms => {
+									handleSearchTerms( terms );
```
Same situation as the other callbacks, wrapper arrow function not strictly needed. (It's not hurting anything, but to my eyes passing the handler to the `.then()` directly is a little faster to understand when skimming)

kadamwhite reviewed (changes requested):
One typo to fix, then some optional cleanup around removing unneeded wrapping arrow functions and adding JSDoc.
